Pack multiple meshes into vertex and index buffers.

pcwalton opened the pull request:
The underlying allocation algorithm is [`offset-allocator`], which is a port of [Sebastian Aaltonen's `OffsetAllocator`]. It's a fast, simple hard real time allocator in the two-level segregated fit family.

Allocations are divided into two categories: *regular* and *large*. Regular allocations go into one of the shared slabs managed by an allocator. Large allocations get their own individual slabs. Due to platform limitations, on WebGL 2 all vertex buffers are considered large allocations that get their own slabs; however, index buffers can still be packed together. The slab size is 32 MB by default, but the developer can adjust it manually.

The mesh bin key and compare data have been reworked so that the slab IDs are compared first. That way, meshes in the same vertex and index buffers tend to be drawn together. Note that this only works well for opaque meshes; transparent meshes must be sorted into draw order, so there's less opportunity for grouping.

The purpose of packing meshes together is to reduce the number of times vertex and index buffers have to be re-bound, which is expensive. In the future, we'd like to use *multi-draw*, which allows us to draw multiple meshes with a single drawcall, as long as they're in the same buffers. Thus, this patch paves the way toward multi-draw, and with it a GPU-driven pipeline.

Even without multi-draw, this patch results in significant performance improvements. For me, the command submission time (i.e. GPU time plus driver and `wgpu` overhead) for Bistro goes from 4.07ms to 1.42ms without shadows (2.8x speedup); with shadows it goes from 6.91ms to 2.62ms (2.45x speedup). The number of vertex and index buffer switches in Bistro is reduced from approximately 3,600 to 927, with the vast majority of the remaining switches due to the transparent pass.

[`offset-allocator`]: https://github.com/pcwalton/offset-allocator/

[Sebastian Aaltonen's `OffsetAllocator`]: https://github.com/sebbbi/OffsetAllocator/

Bistro, without shadows. Yellow is this PR; red is `main`.
![Screenshot 2024-05-02 200257](https://github.com/bevyengine/bevy/assets/157897/ab8da9b4-3c48-4053-a907-647bd970090f)

Bistro, with shadows. Yellow is this PR; red is `main`.
![Screenshot 2024-05-02 200413](https://github.com/bevyengine/bevy/assets/157897/99295514-13eb-4024-b81d-675893b52c3b)

---

## Changelog

### Added
* Multiple meshes can now be packed together into vertex and index buffers, which reduces state changes and provides performance improvements.

## Migration Guide
* The vertex and index data in `GpuMesh` are now `GpuAllocation`s instead of `Buffer`s, to facilitate packing multiple meshes in the same buffer. To fetch the buffer corresponding to a `GpuAllocation`, use the `buffer()` method in the new `GpuAllocator` resource. Note that the allocation may be located anywhere in the buffer; use the `offset()` method to determine its location.

NthTensor commented:
Awesome! I will try to read through and test this over the weekend. 

Will this also give us a second perf improvement when order-independent-transparency lands? That should mean we can basically stop sorting, right?

pcwalton commented:
@NthTensor Yes, I would think so.

Swoorup commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,563 @@
+/// position*, we must only group meshes with identical vertex buffer layouts
+/// into the same buffer.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum GpuAllocationClass {
```
Curious if this makes sense to add other `GpuAllocationClass` textures, samplers, pipeline layout, etc?

pcwalton replied:
I think we should add those as we come to them. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) :)

NthTensor commented on crates/bevy_core_pipeline/src/core_3d/mod.rs:
```diff
@@ -191,22 +193,57 @@ pub struct Opaque3d {
+    /// The mesh.
+    ///
+    /// Although we don't have multidraw capability yet, we place this at the
+    /// end to maximize multidraw opportunities in the future.
```
Why does order matter here? It's not repr(c) or anything.

pcwalton replied:
`PartialOrd` compares from first struct field to last struct field. So fields at the top of the struct will generally be placed together more often.

NthTensor replied:
TIL, Thanks.

NthTensor reviewed (approved):
Some of the code involving the allocator itself is a bit beyond me, but I don't see any issues reading it over. Very pleased with the performance trying this out on my fairly low quality hardware.

re0312 commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,563 @@
+}
+
+/// Manages allocations in GPU buffers.
+#[derive(Resource, Clone)]
```
Seems no need to derive Clone here.

re0312 reviewed (commented):
Outstanding work ! , but looks like something was broken in `2d_shapes`
![image](https://github.com/bevyengine/bevy/assets/45868716/5174ca1b-63fe-4615-a7a5-82ff9db5635f)

pcwalton commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,563 @@
+impl GpuAllocator {
+    /// Returns the slab that the given allocation is stored in.
+    pub fn buffer(&self, allocation: &GpuAllocation) -> &Buffer {
+        &self.slabs[&allocation.slab_id]
```
If it fails it always indicates a bug, because we should be spilling into large allocations if a small allocation can't handle it.

pcwalton commented:
Updated to main and fixed the 2D meshes problem. It was a simple mistake when porting the logic from 3D over: in the indexed path, for the `draw_indexed` method call I forgot to switch the `base_vertex` parameter from 0 to the actual location in the buffer.

atlv24 commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,563 @@
+    /// This exists because copies in `wgpu` must begin and end on 4-byte
+    /// boundaries, so we have to pad out allocations appropriately.
+    fn aligned_unit_size(&self) -> u32 {
+        let mut unit_size = self.unit_size();
+        if unit_size % 4 != 0 {
+            unit_size += 4 - unit_size % 4;
+        }
+        unit_size
```
this could also just be done by masking since its a power of two: `(x + 3) & ~3`  (it probably compiles to this anyways)

pcwalton commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,563 @@
+        let mut slab_id = None;
+        for slab_index in 0..class_allocator_data.free_large_slabs.len() {
+            if self.slabs[&class_allocator_data.free_large_slabs[slab_index]].size()
+                >= contents.len() as u64 * 4
```
Good catch, fixed. That was from an earlier version of the patch where `contents` was `&[u32]` and I forgot to update it.

pcwalton commented:
Comments addressed

alice-i-cecile commented:
> [2:32 PM]pcwalton: I think 0.15 would be best for that.
> [2:32 PM]pcwalton: Specifically my concern is the size of the slabs: the heuristic hasn't been well tuned and might balloon memory usage in some cases. We won't know without testing.
> [2:32 PM]pcwalton: It's the kind of thing we'll only know through a cycle of testing on main so I'd be uncomfortable with 0.14. Besides, @Griffin found that the savings are very situational 

Blocking until 0.14 is shipped.

pcwalton commented:
This is ready to go, but I think it would be best to wait until 0.15 and not merge for 0.14. The reason is that the memory usage heuristics haven't been well tuned yet. We'll only know what the best heuristics are through a cycle of testing.

BD103 commented:
I would like to nominate this for the release notes. I think the performance gains are significant enough that users would enjoy reading about it. (Not to mention I've been watching this in the background every since `offset-allocator` hit the front page of HN!)

alice-i-cecile commented:
Agreed :) In the future, feel free to just add the label yourselves: it's easy to make the editorial call to split or lump things during the final release notes process.

Elabajaba commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,560 @@
+const SWEEP_INTERVAL: Duration = Duration::from_secs(10);
+
+/// The default size of a slab, in bytes.
+const DEFAULT_SLAB_SIZE: u64 = 32 * 1024 * 1024;
```
Should this (in the future) be synced up with wgpu's allocation sizes as well? (https://github.com/gfx-rs/wgpu/pull/5875)

Elabajaba commented on crates/bevy_render/src/allocator.rs:
```diff
@@ -0,0 +1,560 @@
+        class: &GpuAllocationClass,
+        contents: &[u8],
+    ) -> GpuAllocation {
+        // If this is going to overflow a slab, give it
```
```suggestion
        // If this is going to overflow a slab, give it its own large slab.
```

pcwalton commented:
I'm closing this because I've written it.
